### PR TITLE
try to avoid DDL issue

### PR DESCRIPTION
**Pull request by henter**
碰到的问题：
添加\删除字段后无法继续同步，只能重新 dump 重新同步，成本太高。

尝试处理为：
碰到 DDL 时重新初始化对应的 table 信息，刷新内存里的字段信息

注：需在 rule 配置里指定主键，避免因`getDocID`里调用`GetPKValues`时检查字段数量而报错
https://github.com/siddontang/go-mysql-elasticsearch/blob/master/river/sync.go#L405

```
[[rule]]
schema = "mydb"
table = "test"
index = "local_db_test"
type = "test"
id = ["id"]  #指定主键
```

结果：
删除字段后，es 里不会对应删除数据
新增字段后，es 里对应的新字段无 mapping 信息

另外，下面的操作场景会出现脏数据，请注意：
```
停止同步
写入（旧数据）
新增字段（位置非尾部）
写入（新数据）
启动同步
```

### CODE REVIEW

File: river/sync.go [hunk 1]
```diff
@@ -50,7 +50,15 @@ func (h *eventHandler) OnRotate(e *replication.RotateEvent) error {
 	return h.r.ctx.Err()
 }
 
-func (h *eventHandler) OnDDL(nextPos mysql.Position, _ *replication.QueryEvent) error {
+func (h *eventHandler) OnDDL(nextPos mysql.Position, e *replication.QueryEvent) error {
+	log.Info("re-prepare rule info")
+	h.r.c.Sources = nil
```
Comment (henter) [inline river/sync.go]: 这里将 `h.r.c.Sources` 设为 nil 是为了避免 `parseSource` 方法里的报错
https://github.com/siddontang/go-mysql-elasticsearch/blob/master/river/river.go#L153

File: river/sync.go [hunk 2]
```diff
@@ -350,6 +358,12 @@ func (r *River) makeInsertReqData(req *elastic.BulkRequest, rule *Rule, values [
 		if !rule.CheckFilter(c.Name) {
 			continue
 		}
+
+		//if added column ? (not safe, use your own risk)
```
Comment (henter) [inline river/sync.go]: 避免因新增字段导致溢出

**Comment by henter** [comment]
#164 #136

File: river/sync.go [hunk 3]
```diff
@@ -50,7 +50,15 @@ func (h *eventHandler) OnRotate(e *replication.RotateEvent) error {
 	return h.r.ctx.Err()
 }
 
-func (h *eventHandler) OnDDL(nextPos mysql.Position, _ *replication.QueryEvent) error {
+func (h *eventHandler) OnDDL(nextPos mysql.Position, e *replication.QueryEvent) error {
+	log.Info("re-prepare rule info")
+	h.r.c.Sources = nil
+	err := h.r.prepareRule()
```
Comment (siddontang) [inline river/sync.go]: I think you just need to re-prepare the rule for the changing table.

File: river/sync.go [hunk 4]
```diff
@@ -50,7 +50,15 @@ func (h *eventHandler) OnRotate(e *replication.RotateEvent) error {
...
+	if err != nil {
+		log.Errorf("re-prepare rule error %s\n", err.Error())
+		return err
+	}
+
```
Comment (siddontang) [inline river/sync.go]: using err is enough, no need .Error here

File: river/sync.go [hunk 5]
```diff
@@ -50,7 +50,15 @@ func (h *eventHandler) OnRotate(e *replication.RotateEvent) error {
 	return h.r.ctx.Err()
 }
 
-func (h *eventHandler) OnDDL(nextPos mysql.Position, _ *replication.QueryEvent) error {
+func (h *eventHandler) OnDDL(nextPos mysql.Position, e *replication.QueryEvent) error {
+	log.Info("re-prepare rule info")
```
Comment (siddontang) [inline river/sync.go]: please output the DDL here.

**Comment by henter** [comment]
@siddontang Thanks.
just copped func `checkRenameTable` from `siddontang/go-mysql/canal/sync.go` to get table info from DDL

File: river/sync.go [hunk 6]
```diff
@@ -50,7 +51,19 @@ func (h *eventHandler) OnRotate(e *replication.RotateEvent) error {
 	return h.r.ctx.Err()
 }
 
-func (h *eventHandler) OnDDL(nextPos mysql.Position, _ *replication.QueryEvent) error {
+func (h *eventHandler) OnDDL(nextPos mysql.Position, e *replication.QueryEvent) error {
+	mb := checkRenameTable(e)
```
Comment (siddontang) [inline river/sync.go]: seem this is only for rename table DDL, not for others like add/del columns.

File: river/sync.go (hunk 2 above)
Comment (siddontang) [reply river/sync.go]: how about removing column?

File: river/sync.go (hunk 6 above)
Comment (henter) [reply river/sync.go]: also works for  alter table ddl
maybe need to change the func name ? :-)

Comment (henter) [reply river/sync.go]: `checkRenameTable `  ->  `getTableFromDDL `

**Comment by siddontang** [comment]
Hi @henter 

I think this PR is not enough and we should handle more of the DDL. We should handle this in go-mysql too, not only here. 

I will think a workable way then discuss with you later.

**Comment by henter** [comment]
ok, thanks @siddontang

**Pull request closed without merging**